src/kernels/kernel.base.ts

```diff
@@ -191,12 +190,12 @@ export abstract class BaseKernel implements IKernel {
     public get pendingCells(): readonly NotebookCell[] {
         return this.kernelExecution.queue;
     }
-    public async executeCell(cell: NotebookCell): Promise<NotebookCellRunState> {
+    public async executeCell(cell: NotebookCell, controller: NotebookController): Promise<NotebookCellRunState> {
```

Comment:
Not 100% sure but I don't think I found where this is called without the 'this.controller'? Why did you need this extra parameter?

Reply:
In kernel we automatically assume the controller to be used is the same controller associated with the kernel.
E.g. asume we have a notebook A that starts kernel K1 associated with Controller C1.
Now we have a like kernel and we create conroller C2, and we don't really ever spin up a new kernel IKernel as we have one K1

Next, when run run a cell againts notebook B using controller C2, we need to use kernel K1, however kernel K1 will use its own controller C1 to udpate the cell outputs and the like. However thats wrong. We hould be using controller C2 to update contents/output of notebook B.

Hence, when calling executeCell in kernel K1, we need to pass in the controller.
This way a kernel can be use against multiple controllers